Docs: Add MongoDB & ClickHouse tutorials (#796)
hasura/ddn-docs: docs/how-to-build-with-ddn/with-mongodb.mdx

services:
  mongodb:
    image: mongo:latest
    container_name: mongodb
    ports:
      - "27017:27017"
```

```sh title="Run the container:"
docker compose -f app/connector/my_mongo/compose.mongo.yaml up -d
```

```sh title="Use the mongosh shell to seed the database:"
docker exec -it mongodb mongosh my_database --eval "
db.users.insertMany([
  { user_id: 1, name: 'Alice', age: 25 },
  { user_id: 2, name: 'Bob', age: 30 },
  { user_id: 3, name: 'Charlie', age: 35 }
]);
"
```

The shell will return information about the newly-inserted `users` records.

### Step 6. Introspect your MongoDB database

```sh title="Next, use the CLI to introspect your MongoDB database:"
ddn connector introspect my_mongo
```

After running this, you should see a representation of your collection's schema in
`app/connector/my_mongo/schema/users.json`; you can view this using `cat` or open the file in your editor.

For each collection in your database, the MongoDB connector will generate a separate JSON file representing it.

### Step 7. Add your model

```sh title="Now, track the collection from your MongoDB database as a model:"
ddn models add my_mongo users

> ```

Author: ✅ Fixed in PG as well.

> Open the `app/metadata` directory and you'll find a newly-generated file: `Users.hml`. The Hasura engine will use this
> Hasura Metadata Language file to bring the `users` table from MongoDB into your API as a
> [model](/supergraph-modeling/models.mdx).

Reviewer: > The Hasura engine will use this Hasura Metadata Language file to bring the `users` table from MongoDB into your API as a [model](/supergraph-modeling/models.mdx).

This doesn't feel accurate enough. The *DDN CLI* takes hml metadata files with objects and creates a big JSON blobs (builds) for the Hasura engine to run as an API. 

If we're wanting to say *what happened* then the DDN CLI took the introspection of the collection and created an HML model representation for it in the supergraph. 

If we're wanting to say *what's about to happen* then the DDN CLI will take the hml model representation of the mongo collection and create a JSON blob build for the Hasura engine to run. (which also contains models etc) 

If it's what's about to happen my suggestion would be something like:

```suggestion
Open the `app/metadata` directory and you'll find a newly-generated file: `Users.hml`. The DDN CLI will now use this Hasura Metadata Language file to represent the `users` collection from MongoDB in your API as a
[model](/supergraph-modeling/models.mdx).
```

But could easily be what happened too, not what's about to happen. 

Also table == collection for mongo

Noting that we have the same kind of issue with the postgres version too.

Author: ✅ Fixed in PG, too.

  { user_id: 2, title: 'Bob\'s Post', content: 'Bob shares his thoughts.' },
  { user_id: 3, title: 'Hello World', content: 'Charlie joins the conversation.' }
]);
"
```

### Step 12. Refresh your metadata and rebuild your project

:::tip

The following steps are necessary each time you make changes to your **source** schema. This includes, adding,
modifying, or dropping collections.

:::

#### Step 12.1. Re-introspect your data source

```sh title="Run the introspection command again:"
ddn connector introspect my_mongo
```

In `app/connector/my_mongo/configuration.json`, you'll see schema updated to include operations for the `posts`
collection. You'll also see a `posts.json` file in the `schema` directory.

#### Step 12.2. Update your metadata

```sh title="Add the posts model:"
ddn model add my_mongo posts
```

#### Step 12.3. Kill your services

Bring down the services by pressing `CTRL+C` in the terminal tab logging their activity.

#### Step 12.4. Create a new build

```sh title="Next, create a new build:"
ddn supergraph build local
```

#### Step 12.5 Restart your services

```sh title="Bring everything back up:"
ddn run docker-start

> ```

Reviewer: Once we get a fix for the error above could we do this one like the postgres flow in two steps of 1: build 2: stop and restart

      }
    ]
  }
}
```

### Step 14. Create a relationship

```yaml title="Open the Posts.hml file and add the following to the end:"
---
kind: Relationship
version: v1
definition:
  name: user
  sourceType: Posts
  target:
    model:
      name: Users
      relationshipType: Object
  mapping:
    - source:
        fieldPath:
          - fieldName: userId
      target:
        modelField:
          - fieldName: userId
```

:::tip LSP-Assisted authoring is available

We've created an extension for VS Code that leverages LSP to make authoring these metadata objects easier. Check it out
[here](https://marketplace.visualstudio.com/items?itemName=HasuraHQ.hasura).

:::

### Step 15. Rebuild your project

Bring down the services by pressing `CTRL+C` in the terminal tab logging their activity.

```sh title="As your metadata has changed, create a new build:"
ddn supergraph build local
```

```sh title="Bring everything back up:"
ddn run docker-start
```

### Step 16. Query using your relationship

```graphql title="Now, execute a nested query using your relationship:"
query GetPosts {
  posts {
    title
    content
    user {
      userId
      name
      age
    }
  }
}

> ```

Reviewer: Even with image delete and rebuild I'm getting 

```
app_my_mongo-1    | {"timestamp":"2024-12-04T13:16:10.646248Z","level":"INFO","fields":{"message":"Waiting for suitable server to become available","topologyId":"675055ff052b0c17ff39fe22","operation":"aggregate","selector":"ReadPreference { Mode: Primary }","topologyDescription":"{ Type: Unknown, Servers: [ { Address: local.hasura.dev:27017, Type: Unknown, Error: Kind: I/O error: Connection refused (os error 111), labels: {} } ] }","remainingTimeMS":"29999"},"target":"mongodb::server_selection","span":{"internal.visibility":"user","name":"MongoDB Aggregate Command"},"spans":[{"method":"POST","uri":"/query","version":"HTTP/1.1","name":"request"},{"internal.visibility":"user","name":"/query"},{"internal.visibility":"user","name":"Execute Query Pipeline"},{"internal.visibility":"user","name":"MongoDB Aggregate Command"}]}
app_my_mongo-1    | {"timestamp":"2024-12-04T13:16:18.802281Z","level":"ERROR","fields":{"error":"400 Bad Request: Kind: Server selection timeout: No available servers. Topology: { Type: Unknown, Servers: [ { Address: local.hasura.dev:27017, Type: Unknown, Error: Kind: I/O error: Connection refused (os error 111), labels: {} } ] }, labels: {}\n(details: {})"},"target":"mongodb_connector::mongo_connector","span":{"internal.visibility":"user","name":"/query"},"spans":[{"method":"POST","uri":"/query","version":"HTTP/1.1","name":"request"},{"internal.visibility":"user","name":"/query"}]}
```

With no error in Console, just empty response pane. 

🤷

Reviewer: Oh wait I killed my mongo db. Ignore.

```json title="Which should return a result like this:"
{
  "data": {
    "posts": [
      {
        "title": "My First Post",
        "content": "This is Alice's first post.",
        "user": {
          "userId": 1,
          "name": "Alice",
          "age": 25
        }
      },
      {
        "title": "Another Post",
        "content": "Alice writes again!",
        "user": {
          "userId": 1,
          "name": "Alice",
          "age": 25
        }
      },
      {
        "title": "Bob's Post",
        "content": "Bob shares his thoughts.",
        "user": {
          "userId": 2,
          "name": "Bob",
          "age": 30
        }
      },
      {
        "title": "Hello World",
        "content": "Charlie joins the conversation.",
        "user": {
          "userId": 3,
          "name": "Charlie",
          "age": 35
        }
      }
    ]
  }
}
```

## Next steps

Congratulations on completing your first Hasura DDN project with MongoDB! 🎉

Here's what you just accomplished:

- You started with a fresh project and connected it to a local MongoDB database.

> - You set up metadata to represent your tables and relationships, which acts as the blueprint for your API.

Reviewer: ```suggestion
- You set up metadata to represent your collections and relationships, which acts as the blueprint for your API.
```

> - Then, you created a build—essentially compiling everything into a ready-to-use API—and successfully ran your first
>   GraphQL queries to fetch data.

Reviewer: ```suggestion
- Then, you created a build - essentially compiling everything into a ready-to-use API - and successfully ran your first
  GraphQL queries to fetch data.
```

### Step 2. Scaffold out a new local project

```sh title="Next, create a new local project:"
ddn supergraph init my-project && cd my-project
```

Once you move into this directory, you'll see your project scaffolded out for you. You can view the structure by either
running `ls` in your terminal, or by opening the directory in your preferred editor.

### Step 3. Initialize your MongoDB connector

```sh title="In your project directory, run:"
ddn connector init my_mongo -i
```

From the dropdown, start typing `mongo` and hit enter to accept the default port. Then, provide the following connection
string:

```plaintext
mongodb://local.hasura.dev:27017/my_database
```

### Step 4. Start the local MongoDB container

```sh title="Begin by creating a compose file for the Mongo service:"
touch app/connector/my_mongo/compose.mongo.yaml
```

```yaml title="Then, open the file and add the following:"
services:
  mongodb:
    image: mongo:latest
    container_name: mongodb
    ports:
      - "27017:27017"
```

```sh title="Run the container:"
docker compose -f app/connector/my_mongo/compose.mongo.yaml up -d
```

```sh title="Use the mongosh shell to seed the database:"

> docker exec -it mongodb mongosh --eval "
> db = db.getSiblingDB('my_database');
> db.users.insertMany([
>   { user_id: 1, name: 'Alice', age: 25 },
>   { user_id: 2, name: 'Bob', age: 30 },
>   { user_id: 3, name: 'Charlie', age: 35 }
> ]);
> "

Reviewer 2: Another way to select the database is to give it as an argument to the mongosh command:

```suggestion
docker exec -it mongodb mongosh my_database --eval "
db.users.insertMany([
  { user_id: 1, name: 'Alice', age: 25 },
  { user_id: 2, name: 'Bob', age: 30 },
  { user_id: 3, name: 'Charlie', age: 35 }
]);
"
```

It's a little more concise, although possibly it's easier to overlook the important database argument.

```

The shell will return information about the newly-inserted `users` records.

### Step 6. Introspect your MongoDB database

```sh title="Next, use the CLI to introspect your MongoDB database:"
ddn connector introspect my_mongo
```

After running this, you should see a representation of your collection's schema in
`app/connector/my_mongo/schema/users.json`; you can view this using `cat` or open the file in your editor.

For each collection in your database, the MongoDB connector will generate a separate JSON file representing it.

> ### Step 7. Add your model

Reviewer 2: For now I think we need to restart the connector between steps 6 & 7. This shouldn't be necessary after we implement the "print schema" capability in the mongo connector. I've added a subtask to ENG-1270 to update this tutorial once that is implemented.

Author: Here, we don't have the connector running yet.

Reviewer 2: Oh right - good point

        "age": 30
      },
      {
        "userId": 3,
        "name": "Charlie",
        "age": 35
      }
    ]
  }
}
```

### Step 11. Iterate on your MongoDB schema

```sh title="Let's add a new collection for posts:"

docker exec -it mongodb mongosh --eval "
db = db.getSiblingDB('my_database');
db.posts.insertMany([
  { user_id: 1, title: 'My First Post', content: 'This is Alice\'s first post.' },
  { user_id: 1, title: 'Another Post', content: 'Alice writes again!' },
  { user_id: 2, title: 'Bob\'s Post', content: 'Bob shares his thoughts.' },
  { user_id: 3, title: 'Hello World', content: 'Charlie joins the conversation.' }
]);
"
```

### Step 12. Refresh your metadata and rebuild your project

:::tip

The following steps are necessary each time you make changes to your **source** schema. This includes, adding,
modifying, or dropping collections.

:::

#### Step 12.1. Re-introspect your data source

```sh title="Run the introspection command again:"
ddn connector introspect my_mongo
```

In `app/connector/my_mongo/configuration.json`, you'll see schema updated to include operations for the `posts`
collection. You'll also see a `posts.json` file in the `schema` directory.

> #### Step 12.2. Update your metadata

Reviewer 2: Same deal, I think we need a connector restart between steps 12.1 and 12.2 until the "print schema" capability is implemented

Author: Is it possible there's some caching of a docker build going on? Despite adding this step in, I'm still having to manually delete the image and then rerun `ddn run docker-start` to get my changes to appear in the connector.

Reviewer: Same. I get this error: 

<img width="1164" alt="Screenshot 2024-12-04 at 13 10 21" src="https://github.com/user-attachments/assets/c10e6669-1dd4-4ce6-8a38-99c8b6aa80c7">
